nilearn/regions/tests/test_signal_extraction.py

```diff
@@ -17,14 +17,65 @@
 from nilearn._utils.exceptions import DimensionError
 from nilearn.image import get_data, new_img_like
 
+
 _TEST_DIM_ERROR_MSG = ("Input data has incompatible dimensionality: "
                        "Expected dimension is 3D and you provided "
                        "a 4D image")
+_TEST_SHAPE_ERROR_MSG = "Images have incompatible shapes."
+_TEST_AFFINE_ERROR_MSG = "Images have different affine matrices."
 
+INF = 1000 * np.finfo(np.float32).eps
 
-def test_generate_regions_ts():
-    """Minimal testing of generate_regions_ts()"""
+        
+def test__check_shape_and_affine_compatibility_without_dim():
+    """Ensure correct behaviour for valid data without dim"""
+    shape = (2, 3, 4)
+    affine = np.eye(4)
+    img1 = nibabel.Nifti1Image(np.zeros(shape), affine)
+
+    img2 = nibabel.Nifti1Image(np.zeros(shape), affine)
+
+    with pytest.warns(None):
+        signal_extraction._check_shape_and_affine_compatibility(
+            img1,
+            img2)
+        
+        
+def test__check_shape_and_affine_compatibility_with_dim():
+    """Ensure correct behaviour for valid data without dim"""
+    shape = (2, 3, 4, 7)
+    affine = np.eye(4)
+    img1 = nibabel.Nifti1Image(np.zeros(shape[:3]), affine)
+
+    mask_shape = (2, 3, 4)
+    img2 = nibabel.Nifti1Image(np.zeros(mask_shape), affine)
+
+    with pytest.warns(None):
+        signal_extraction._check_shape_and_affine_compatibility(
+            img1,
+            img2,
+            dim=3)
+
+@pytest.mark.parametrize(
+    "test_shape,test_affine,msg", 
+    [((2, 3, 5), np.eye(4), _TEST_SHAPE_ERROR_MSG),
+     ((2, 3, 4), 2*np.eye(4), _TEST_AFFINE_ERROR_MSG)
+     ])
+def test__check_shape_and_affine_compatibility_error(test_shape,test_affine,msg):
+    shape = (2, 3, 4)
+    affine = np.eye(4)
+    img1 = nibabel.Nifti1Image(np.zeros(shape), affine)
+
+    img2 = nibabel.Nifti1Image(np.zeros(test_shape), test_affine)
+
+    with pytest.raises(ValueError, match=msg):
+        signal_extraction._check_shape_and_affine_compatibility(
+            img1,
+            img2)
```

Comment:
@NicolasGensollen does this look OK to you? both readable and succinct?

Reply:
@Remi-Gau Yes, looks great ! Thanks for the refactoring ! 👍